pdns/dnsdistdist/doh.cc

```diff
@@ -0,0 +1,707 @@
+    h2o_send_error_400(req, "Bad Request", "dnsdist " VERSION " could not parse DNS query", 0);
+  }
+}
+
+/*
+   For GET, the base64url-encoded payload is in the 'dns' parameter, which might be the first parameter, or not.
+   For POST, the payload is the payload.
+ */
+static int doh_handler(h2o_handler_t *self, h2o_req_t *req)
+try
+{
+  // g_logstream<<(void*)req<<" doh_handler"<<endl;
+  if(!req->conn->ctx->storage.size) {
+    return 0; // although we might was well crash on this
+  }
+  h2o_socket_t* sock = req->conn->callbacks->get_socket(req->conn);
+  ComboAddress remote;
+  h2o_socket_getpeername(sock, reinterpret_cast<struct sockaddr*>(&remote));
+  DOHServerConfig* dsc = reinterpret_cast<DOHServerConfig*>(req->conn->ctx->storage.entries[0].data);
+
+  if(auto tlsversion = h2o_socket_get_ssl_protocol_version(sock)) {
+    if(!strcmp(tlsversion, "TLSv1.0"))
+      ++dsc->df->d_tls10queries;
+    else if(!strcmp(tlsversion, "TLSv1.1"))
+      ++dsc->df->d_tls11queries;
+    else if(!strcmp(tlsversion, "TLSv1.2"))
+      ++dsc->df->d_tls12queries;
+    else if(!strcmp(tlsversion, "TLSv1.3"))
+      ++dsc->df->d_tls13queries;
+    else
+      ++dsc->df->d_tlsUnknownqueries;
+  }
+
+  string path(req->path.base, req->path.len);
+
+  if (h2o_memis(req->method.base, req->method.len, H2O_STRLIT("POST"))) {
+    ++dsc->df->d_postqueries;
+    if(req->version >= 0x0200)
+      ++dsc->df->d_http2queries;
+    else
+      ++dsc->df->d_http1queries;
+
+    std::string query;
+    query.reserve(req->entity.len + 512);
+    query.assign(req->entity.base, req->entity.len);
+    doh_dispatch_query(dsc, self, req, std::move(query), remote);
+  }
+  else if(req->query_at != SIZE_MAX && (req->path.len - req->query_at > 5)) {
+    auto pos = path.find("?dns=");
+    if(pos == string::npos)
+      pos = path.find("&dns=");
+    if(pos != string::npos) {
+      // need to base64url decode this
+      string sdns(path.substr(pos+5));
+      boost::replace_all(sdns,"-", "+");
+      boost::replace_all(sdns,"_", "/");
+      sdns.append(sdns.size() % 4, '='); // re-add padding that may have been missing
+
+      string decoded;
+      /* rough estimate so we hopefully don't need a need allocation later */
+      decoded.reserve(((sdns.size() * 3) / 4) + 512);
+      if(B64Decode(sdns, decoded) < 0) {
+        h2o_send_error_400(req, "Bad Request", "dnsdist " VERSION " could not decode BASE64-URL", 0);
+        ++dsc->df->d_badrequests;
+        return 0;
+      }
+      else {
+        ++dsc->df->d_getqueries;
+        if(req->version >= 0x0200)
+          ++dsc->df->d_http2queries;
+        else
+          ++dsc->df->d_http1queries;
+
+        doh_dispatch_query(dsc, self, req, std::move(decoded), remote);
+      }
+    }
+    else
+    {
+      vinfolog("HTTP request without DNS parameter: %s", req->path.base);
+      h2o_send_error_400(req, "Bad Request", "dnsdist " VERSION " could not find DNS parameter", 0);
```

Comment:
I see that errors - mostly - send the dnsdist's version in their explanation message. Maybe this would be nicer to add a HTTP header (such as `x-dnsdist-version`) with dnsdist version and and option to disable that ?

Reply:
I really don't think we should advertise the product and version at all. Unless someone strongly disagrees, I'll remove "dnsdist' and the version from these messages.

Reply:
If we need to make the choice, it's better not to advertise. `version.bind` and friends are configurable in auth and rec, I suppose that would be the obvious thing to do -eventually-.

Reply:
On the same topic, h2o advertises it's version "Server: h2o/2.2.5". Maybe you could also remove it :)